### ISSUE TITLE
Incorrect smart indent anchor following `do` keyword

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn#50063]
**Version Used**: 16.8.3

🔗 Originally filed as AB#1259025

**Steps to Reproduce**:

```csharp
class Test
{
    void Method()
    {
do$$
    }
}
```

Press <kbd>Enter</kbd> after the `do` keyword.

**Expected Behavior**:

```csharp
class Test
{
    void Method()
    {
do
    $$
    }
}
```

**Actual Behavior**:

```csharp
class Test
{
    void Method()
    {
do
            $$
    }
}
```

**Comment by mmundy3832** [comment]
This auto-indent issue specifically calls out the “do” operator as the anchor. I experience this indentation issue for all “anchors”.

See attached video where I demonstrate the same issue with for(), if(), and a function declaration.

https://user-images.githubusercontent.com/17011063/103379340-e746ba80-4aaa-11eb-806f-ae7b40128cc2.mp4

- This was addressed for Java editors in https://github.com/Microsoft/TypeScript/issues/21139, but this same “feature” needs to be added for the C# editors as well. Maybe even all of the editors.

- I would also like an option to create the close brace on the next line if I’m opening the { on an empty line.

This issue was also in VS 2017, and It’s bothered me for quite some time. But now that I know it’s been fixed, at least for Java, it should absolutely be fixed for all other editor formats as well.

**Comment by sharwell** [comment]
> I would also like an option to create the close brace on the next line if I’m opening the { on an empty line.

Can you file a separate issue for this?

**Comment by sharwell** [comment]
@CyrusNajmabadi this seems to be caused by the following:

https://github.com/dotnet/roslyn/blob/c49c9bb5d8c2d34f6d6d986b8c4ed4620d958e92/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/BottomUpBaseIndentationFinder.cs#L98-L99

These parent indent operations assume that the `do` keyword is getting indented, but it's not. We only want to apply the _relative_ indentation.

**Comment by mmundy3832** [comment]
> > I demonstrate the same issue with for(), if(), and a function declaration
> 
> It seems straightforward to expand this issue to these three new cases, based on their similarity to the initial description:
> 
> * `for`
> * `if`
> * Local function declarations
> 
> > I would also like an option to create the close brace on the next line if I’m opening the { on an empty line.
> 
> Can you file a separate issue for this?

I want to be sure that you understand, this happens for virtually all lines that I enter. Not just these specific cases. 

I prefer to butt my code up to the left margin when making changes because it is very easy to identify what I've changed when it's time to clean up and check in my changes. Any time I hit return, and auto format is off, it should not indent the next line that I'm going to type on.

**Comment by sharwell** [comment]
> I want to be sure that you understand, this happens for virtually all lines that I enter. Not just these specific cases.

This is an _extremely_ niche scenario (exactly one known user is impacted). I'm happy to file issues and look into fixing cases that are clearly marked out, but there are rapidly diminishing returns for trying to find them all manually. 😄 

> Any time I hit return, and auto format is off, it should not indent the next line that I'm going to type on.

This is controlled by the indentation settings. If you turn off smart indent, it will stop trying to indent based on formatting context.

**Comment by sharwell** [comment]
📝 A failing test was added for this in #50184. When this issue is fixed, the test can be enabled and used to validate the fix.

**Comment by Dean-NC** [comment]
I filed a Visual Studio bug [report](https://developercommunity.visualstudio.com/t/Too-much-auto-indent-within-using-block/1361556) about wrong indentation when pressing the enter-key for a try/catch, when it's inside a "using block", if you did not use curly braces for the using block.  The VS team said it was a duplicate of this github issue, but I don't think it is.

Please let me know if it's related.

Here's the video I attached to that bug report, that shows the problem: 
[VS-WrongIndent.zip](https://github.com/dotnet/roslyn/files/6123948/VS-WrongIndent.zip)

**Comment by sharwell** [comment]
@Dean-NC your specific case involved an incorrect indent anchor for the `try` keyword. The formatter expected the `try` keyword to be indented one level more than it was, and is indenting all of the contents of that `try` statement one additional level.

Here's the syntax as it appears in the file:

```csharp
using (var stream = new MemoryStream())
try
{
        $$
}
finally
{
}
```

Here's the syntax Visual Studio "wants", and is therefore using for indentation calculation:

```csharp
using (var stream = new MemoryStream())
    try
    {
        $$
    }
    finally
    {
    }
```

**Comment by Dean-NC** [comment]
@sharwell Thank you.  So is my bug report a "duplicate" as the VS team said, or should I ask them to re-open the report since it's not a duplicate of this GitHub issue?

**Comment by sharwell** [comment]
@Dean-NC I would consider your issue a duplicate of this one.

**Comment by Dean-NC** [comment]
@sharwell Yes, I see what you mean...the formatter is not considering the position of the surrounding code.
Thanks again...I'll leave the closed bug report alone, as it should get resolved by the fix in this issue.
